### PR TITLE
[MRG] Adding tests for ExtensionApp

### PR DESCRIPTION
**Pull request by Zsailer**
I started writing the initial unit tests for the `ExtensionApp`. The tests exposed a few ways to better arrange methods in this class. 

The tests are quite simple. No servers are started. This mostly checks that configuration is loaded properly between extension and server.

### CODE REVIEW

**Comment by rolweber** [comment]
Tests are failing. This is from the [Travis job for Python 3.6](https://travis-ci.org/jupyter/jupyter_server/jobs/553531546#L905-L933);
```txt
ERROR: jupyter_server.extension.tests.test_extension.test_prepare_config
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/build/jupyter/jupyter_server/jupyter_server/extension/tests/test_extension.py", line 44, in test_prepare_config
    extension = TestExtensionApp.mock_instance()
  File "/home/travis/build/jupyter/jupyter_server/jupyter_server/extension/tests/test_extension.py", line 16, in mock_instance
    server = self.initialize_server()
  File "/home/travis/build/jupyter/jupyter_server/jupyter_server/extension/application.py", line 224, in initialize_server
    serverapp.initialize(argv=argv)
  File "</home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/decorator.py:decorator-gen-7>", line 2, in initialize
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/traitlets/config/application.py", line 87, in catch_config_error
    return method(app, *args, **kwargs)
  File "/home/travis/build/jupyter/jupyter_server/jupyter_server/serverapp.py", line 1458, in initialize
    super(ServerApp, self).initialize(argv)
  File "</home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/decorator.py:decorator-gen-6>", line 2, in initialize
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/traitlets/config/application.py", line 87, in catch_config_error
    return method(app, *args, **kwargs)
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/jupyter_core/application.py", line 239, in initialize
    self.parse_command_line(argv)
  File "/home/travis/build/jupyter/jupyter_server/jupyter_server/serverapp.py", line 1137, in parse_command_line
    super(ServerApp, self).parse_command_line(argv)
  File "</home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/decorator.py:decorator-gen-4>", line 2, in parse_command_line
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/traitlets/config/application.py", line 93, in catch_config_error
    app.exit(1)
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/traitlets/config/application.py", line 648, in exit
    sys.exit(exit_status)
```

**Comment by Zsailer** [comment]
@rolweber I reworked the tests a bit. Everything passes now. I think this is ready to merge if everything looks good to you. 😄

File: jupyter_server/extension/application.py [hunk 1]
```diff
@@ -214,47 +214,75 @@ def _prepare_templates(self):
...
         # Initialize ServerApp config.
         # Parses the command line looking for 
         # ServerApp configuration.
-        serverapp.initialize()
+        serverapp.initialize(argv=argv)
         return serverapp
 
-    def initialize(self, serverapp, argv=None):
-        """Initialize the extension app."""
+    def initialize(self, serverapp, argv=[]):
```
Comment (vidartf) [inline jupyter_server/extension/application.py]: Lint: Using a mutable arg as default value. Several other such cases in the PR as well.

Comment (rolweber) [reply jupyter_server/extension/application.py]: Good point. Every call will get the same instance of the empty list, so if somebody modifies the list to be non-empty...

Comment (Zsailer) [reply jupyter_server/extension/application.py]: Hm, right. That could cause some inheritance problems. I'll submit a fix in a later PR.

**Pull request merged**